include/bbp/sonata/output.h

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <algorithm>
+#include <map>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include <bbp/sonata/population.h>
+
+namespace {
+template <class InputIt1,
+          class InputIt2,
+          class OutputIt,
+          class BinaryOperation,
+          class BinaryPredicate>
+OutputIt transform_if(InputIt1 first1,
+                      InputIt1 last1,
+                      InputIt2 first2,
+                      OutputIt d_first,
+                      BinaryOperation binary_op,
+                      BinaryPredicate binary_pred) {
+    while (first1 != last1) {
+        if (binary_pred(*first1, *first2)) {
+            *d_first++ = binary_op(*first1, *first2);
+        }
+        ++first1, ++first2;
+    }
+    return d_first;
+}
+}  // unnamed namespace
+
+namespace bbp {
+namespace sonata {
+/**
+  const SpikeReader file(filename);
+  auto pops = file.getPopulationNames();
+  for (const auto& data: file[pops[0]].get(Selection{12UL, 34UL, 58UL})) {
+      uint64_t node_id;
+      double timestamp;
+      std::tie(node_id, timestamp) = data;
+      std::cout << "[" << timestamp << "] " << node_id << std::endl;
+  }
+*/
+class SONATA_API SpikeReader
+{
+  public:
+    class Population
+    {
+      public:
+        std::vector<std::pair<uint64_t, double>> get() const;
+        std::vector<std::pair<uint64_t, double>> get(Selection node_ids) const;
+        std::vector<std::pair<uint64_t, double>> get(double tstart, double tend) const;
+
+      private:
+        explicit Population(const std::string& filename, const std::string& populationName);
+
+        std::vector<uint64_t> node_ids;
+        std::vector<double> timestamps;
```

Comment:
Maybe having these vectors sorted by increasing timestamp could be beneficial afterwards.
This could allow some optimizations for:
`std::vector<std::pair<uint64_t, double>> get(double tstart, double tend) const;`
And would make the outputs more "readable".

What do you think @matz-e , @mgeplf ?

Reply:
I don't think this should be the case, since the SONATA spike file has this attribute called sorting, that sorts the node_ids/timestamps by_time or by_id
https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#output-file-formats

So I guess the users should be able to read the values in the order they were written.

Reply:
hum ok I see.
So it could be a good Idea to give access to this `sorting` field so the consumer don t have to resort the output again if not needed.

Reply:
Miss this part of the specification, I will try to do something nice with it.

Reply:
`SpikeReaderPopulation` => `SpikePopulation`

Reply:
By the way, what should I do about sorting?

Reply:
I believe nothing! This should come sorted from the writing side. Maybe just check if it is sorted according to the enum value?

Reply:
I think checking if the values are correctly sorted is definitely a good idea. 
I don t know how big the reports will be but thanks to the sorting, you can avoid looping over all gids/timestamps everytime you do a get. But I guess if it s fast enough you can skip it.

Reply:
I find the checking useless, if data are corrupted, it's life...

I will try to make some smart search.

Reply:
I find the checking useless, if data are corrupted, it's life...

I will try to make some smart search.

Reply:
> I find the checking useless, if data are corrupted, it's life...

Agreed - if `sorting` says it's sorted, we have to take that at face value - checking if it's sorted is expensive.